engine/src/main/java/org/terasology/engine/rendering/world/ChunkMeshWorker.java

```diff
@@ -132,24 +150,27 @@ public int numberChunkMeshProcessing() {
         return chunksInProximityOfCamera;
     }
 
+    Flux<Chunk> getCompletedChunks() {
+        return completedChunks;
+    }
 
-    static class MeshGenerator {
-        private final ChunkTessellator chunkTessellator;
-        private final WorldProvider worldProvider;
-
-        MeshGenerator(ChunkTessellator chunkTessellator, WorldProvider worldProvider) {
-            this.chunkTessellator = chunkTessellator;
-            this.worldProvider = worldProvider;
-        }
+    private static Chunk uploadNewMesh(Chunk chunk, ChunkMesh chunkMesh) {
+        chunkMesh.updateMesh();  // Does GL stuff, must be on main thread!
+        chunkMesh.discardData();
+        chunk.setMesh(chunkMesh);
+        return chunk;
+    }
 
-        Optional<Tuple2<Chunk, ChunkMesh>> generate(Chunk chunk) {
+    private static Function<Chunk, Mono<Tuple2<Chunk, ChunkMesh>>> generateMeshFunc(ChunkTessellator chunkTessellator, WorldProvider worldProvider) {
+        return (chunk -> {
+            chunk.setDirty(false);
             ChunkView chunkView = worldProvider.getLocalView(chunk.getPosition());
-            if (chunkView != null && chunkView.isValidView() /* && chunkMeshProcessing.remove(chunk.getPosition()) */) {
+            if (chunkView != null && chunkView.isValidView()) {
```

Comment:
FIXME (or test): omitting the `chunkMeshProcessing` part of this `if` condition changes something about the reliability of `ChunkMeshWorker.remove`, I think. Because now there can be things that were removed and are no longer in `chunkMeshProcessing` list, yet they may still run this code.

Is that harmful? Unsure.